gossipsub.go

```diff
@@ -917,6 +967,27 @@ func (gs *GossipSubRouter) handlePrune(p peer.ID, ctl *pb.ControlMessage) {
 	}
 }
 
+func (gs *GossipSubRouter) handleIDontWant(p peer.ID, ctl *pb.ControlMessage) {
+	if gs.unwanted[p] == nil {
+		gs.unwanted[p] = make(map[checksum]int)
+	}
+
+	// IDONTWANT flood protection
+	gs.peerdontwant[p]++
+	if gs.peerdontwant[p] > gs.params.MaxIDontWantMessages {
+		log.Debugf("IDONWANT: peer %s has advertised too many times (%d) within this heartbeat interval; ignoring", p, gs.peerdontwant[p])
+		return
+	}
+
+	// Remember all the unwanted message ids
+	for _, idontwant := range ctl.GetIdontwant() {
+		for _, mid := range idontwant.GetMessageIDs() {
+			hashed := sha256.Sum256([]byte(mid))
```

Comment:
Can we assume that message IDs are small? Lotus, at least, uses hashes already.

We can also inline if and only if the ID is small enough:

```go
type id struct {
	payload [32]byte
	length uint8
}

func newId(mid string) id {
	var out id
	if len(id) > 32 || len(id) == 0 {
		out.payload = sha256.Sum256([]byte(mid))
	} else {
		out.length = uint8(copy(out.payload, mid))
	}
	return id
}
```

Reply:
That makes sense. Let's do that.